crs_application/src/task_managers/motion_planning_manager.cpp

```diff
@@ -53,37 +68,246 @@ MotionPlanningManager::~MotionPlanningManager()
 
 common::ActionResult MotionPlanningManager::init()
 {
-  RCLCPP_WARN(node_->get_logger(),"%s not implemented yet",__PRETTY_FUNCTION__);
+  using namespace crs_msgs;
+  call_freespace_planning_client_ = node_->create_client<srv::CallFreespaceMotion>(CALL_FREESPACE_MOTION_SERVICE);
+  process_motion_planning_client_ = node_->create_client<crs_msgs::srv::PlanProcessMotions>(PLAN_PROCESS_MOTIONS_SERVICE);
+
+  // checking clients
+  std::vector<rclcpp::ClientBase*> clients = {call_freespace_planning_client_.get(),
+                                              process_motion_planning_client_.get()};
+  if(std::all_of(clients.begin(), clients.end(),[](rclcpp::ClientBase* c){
+    return c->wait_for_service(std::chrono::duration<float>(WAIT_SERVICE_DURATION));
+  }))
+  {
+    RCLCPP_WARN(node_->get_logger(),"%s: One or more services were not found", MANAGER_NAME.c_str());
+    return false;
+  }
+
   return true;
 }
 
 common::ActionResult MotionPlanningManager::configure(const MotionPlanningConfig& config)
 {
-  RCLCPP_WARN(node_->get_logger(),"%s not implemented yet",__PRETTY_FUNCTION__);
+  config_ = std::make_shared<MotionPlanningConfig>(config);
   return true;
 }
 
 common::ActionResult MotionPlanningManager::setInput(const datatypes::ProcessToolpathData& input)
 {
-  RCLCPP_WARN(node_->get_logger(),"%s not implemented yet",__PRETTY_FUNCTION__);
+  if(input.rasters.empty())
+  {
+    RCLCPP_ERROR(node_->get_logger(),"%s found no rasters in the process path", MANAGER_NAME.c_str());
+    return false;
+  }
+  input_ = std::make_shared<datatypes::ProcessToolpathData>(input);
+  return true;
+}
+
+
+common::ActionResult MotionPlanningManager::checkPreReq()
+{
+  if(!config_)
+  {
+    common::ActionResult res = { succeeded :false, err_msg: "No configuration has been provided, can not proceed"};
+    RCLCPP_WARN(node_->get_logger(),"%s %s",MANAGER_NAME.c_str(), res.err_msg.c_str());
+  }
+
+  if(!input_)
+  {
+    common::ActionResult res = { succeeded :false, err_msg: "No input data has been provided, can not proceed"};
+    RCLCPP_WARN(node_->get_logger(),"%s %s",MANAGER_NAME.c_str(), res.err_msg.c_str());
+  }
   return true;
 }
 
 common::ActionResult MotionPlanningManager::splitToolpaths()
 {
-  RCLCPP_WARN(node_->get_logger(),"%s not implemented yet",__PRETTY_FUNCTION__);
+  common::ActionResult res = checkPreReq();
+  if( !res )
+  {
+    return res;
+  }
+
+  // computing split distance
+  const double split_dist = config_->tool_speed * config_->media_change_time;
+
+  // splitting paths
+  std::vector<datatypes::ProcessToolpathData> toolpaths_processes;
+  double current_dist = 0.0;
+  Eigen::Vector3d p1, p2;
+  toolpaths_processes.resize(1);
+  for(std::size_t r_idx = 0 ; r_idx < input_->rasters.size(); r_idx++)
+  {
+    geometry_msgs::msg::PoseArray& raster = input_->rasters[r_idx];
+    int split_point_idx = -1;
+    for(std::size_t p_idx = 0; p_idx < raster.poses.size() - 1; p_idx++)
```

Comment:
I think this fails to account for the remaining distance in a raster if a split occurs and instead will pickup at the beginning of the next raster with a distance of zero.

Reply:
I think it captures the remaining of the raster inside the if statement
```
    if(split_point_idx >=0)
    {
    .
    .
    .
```